connexion/operations/swagger2.py

```diff
@@ -224,30 +234,95 @@ def _nested_example(self, schema):
+    def body_definition(self, content_type: str = None) -> dict:
         """
         The body complete definition for this operation.
 
         **There can be one "body" parameter at most.**
 
         :rtype: dict
         """
-        body_parameters = [p for p in self.parameters if p["in"] == "body"]
-        if len(body_parameters) > 1:
-            raise InvalidSpecification(
-                "{method} {path} There can be one 'body' parameter at most".format(
-                    method=self.method, path=self.path
+        if content_type in FORM_CONTENT_TYPES:
+            form_parameters = [p for p in self.parameters if p["in"] == "formData"]
+            body_definition = self._transform_form(form_parameters)
+        else:
+            body_parameters = [p for p in self.parameters if p["in"] == "body"]
+            if len(body_parameters) > 1:
+                raise InvalidSpecification(
+                    "{method} {path} There can be one 'body' parameter at most".format(
+                        method=self.method, path=self.path
+                    )
                 )
-            )
-        return body_parameters[0] if body_parameters else {}
+            body_definition = body_parameters[0] if body_parameters else {}
+        return body_definition
+
+    def _transform_form(self, form_parameters: t.List[dict]) -> dict:
+        """Translate Swagger2 form parameters into OpenAPI 3 jsonschema spec"""
+        properties = {}
+        required = []
+        collection_formats = set()
+
+        for param in form_parameters:
+            prop = {}
+
+            if param["type"] == "file":
+                prop.update(
+                    {
+                        "type": "string",
+                        "format": "binary",
+                    }
+                )
+            else:
+                prop["type"] = param["type"]
+
+            format_ = param.get("format")
+            if format_ is not None:
+                prop["format"] = format_
+
+            default = param.get("default")
+            if default is not None:
+                prop["default"] = default
+
+            nullable = param.get("x-nullable")
+            if nullable is not None:
+                prop["nullable"] = nullable
+
+            if param["type"] == "array":
+                prop["items"] = param.get("items", {})
+                collection_formats.add(param.get("collectionFormat", "csv"))
+
+            properties[param["name"]] = prop
+
+            if param.get("required", False):
+                required.append(param["name"])
+
+        definition = {
+            "schema": {
+                "type": "object",
+                "properties": properties,
+                "required": required,
+                "additionalProperties": not self.strict_validation,
+            }
+        }
+
+        if collection_formats:
+            if len(collection_formats) > 1:
+                raise InvalidSpecification(
```

Comment:
As far as I'm aware, this and the `tsv` collection format are the only two aspects of the Swagger 2 parameter spec that cannot be translated to the OpenAPI 3 style request body spec.

I think these cases are rare enough that we can decide not to support them so we can unify form validation across versions.

Reply:
I'll add an explicit error for the `tsv` collection format as well though.

Reply:
Indeed, I believe you're right. There are just a few exceptions where Swagger 2 and OpenAPI 3:

> This instance [tsv], and the case where a parameter has a type of array and a collectionFormat of x (e.g. pipes) and an items.collectionFormat of y (e.g. csv) (giving: a,b|c,d) seem to be the only parts of the 2.0 specification which can't be losslessly converted to 3.0.x

https://github.com/OAI/OpenAPI-Specification/issues/1046#issuecomment-297436141

Reply:
Ok, so multiple collection formats for a single operation does work, just not multiple collection formats for the same parameter / property. That makes it even more rare. Updated the code to reflect this.